Adding health check server hook in torch elastic

gag1jain opened the pull request:
Summary:
Building hook for external mechanism to monitor the health of torch elastic launcher. Health check server takes dependency on FileTimerServer to check if launcher is healthy or not. It will be always healthy if FileTimerServer is disabled.

Implementation of start_healthcheck_server is unsupported, however tcp/http server can be started on specific port which can monitor the aliveness of worker_watchdog and accordingly take the action.

Test Plan: buck test mode/opt caffe2/test/distributed/elastic/agent/server/test:local_agent_test

Differential Revision: D55108182




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

kurman reviewed (approved):
stamping to unblock. Please address CI failures on GH before landing.

gag1jain committed:
Adding health check server hook in torch elastic (#122750)

Summary:

Building hook for external mechanism to monitor the health of torch elastic launcher. Health check server takes dependency on FileTimerServer to check if launcher is healthy or not. It will be always healthy if FileTimerServer is disabled.

Implementation of start_healthcheck_server is unsupported, however tcp/http server can be started on specific port which can monitor the aliveness of worker_watchdog and accordingly take the action.

Test Plan: buck test mode/opt caffe2/test/distributed/elastic/agent/server/test:local_agent_test

Reviewed By: kurman

Differential Revision: D55108182

huydhn commented:
@gag1jain  Could you help take a look at the docs build failure on OSS CI https://hud.pytorch.org/pytorch/pytorch/commit/61d431fab07f65d3e54c28f1ec420c517c7ada92? This looks like a land race with https://github.com/pytorch/pytorch/pull/123425 and the failure passed OSS CI without being caught by OSS CI.

The error is
`You added the following module(s) to the PyTorch namespace 'torch.distributed.elastic.agent.server.health_check_server' but they have no corresponding entry in a doc .rst file. You should either make sure that the .rst file that contains the module's documentation properly contains either '.. automodule:: mod_name' (if you do not want the paragraph added by the automodule, you can simply use '.. py:module:: mod_name')  or make the module private (by appending an '_' at the beginning of its name).+ echo 'Make sure you'\''ve updated relevant .rsts in docs/source!'`.

malfet commented:
Unlanding the change as it broke docs build
